### ISSUE TITLE
"MongoError $setOnInsert' is empty" when using immutable createdAt

### ISSUE THREAD

**Issue by andrewpile** [Automattic/mongoose#7917]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

When performing an `update*`, `$setOnInsert` is set to empty if the schema is using the `timestamps` option and the `createdAt` property is marked as `immutable: true`, even if the `update` query specifies `{ upsert: false }`.

**If the current behavior is a bug, please provide the steps to reproduce.**

The model must include the appropriate timestamps and mark `createdAt` as `immutable`:

```
    createdAt: {
        type: mongoose.Schema.Types.Date,
        default: (new Date()).getTime(),
        index: true,
        immutable: true
    },
    updatedAt: {
        type: mongoose.Schema.Types.Date,
        default: (new Date()).getTime()
    }
}, {
    timestamps: true
})
```

Upon attempting `Model.update({ .. }, { upsert: false })`, the Mongo query will be of the form:

`Model.update({ ... }, { '$setOnInsert': {}, '$set': { ... }, { upsert: false })`

This results in an error:

`MongoError: '$setOnInsert' is empty. You must specify a field like so: {$setOnInsert: {<field>: ...}}`

**What is the expected behavior?**

Making `createdAt` immutable is a desirable use case: http://thecodebarbarian.com/whats-new-in-mongoose-5-6-immutable-properties.html

It appears `immutable` properties are being exempted from `$setOnInsert`? At the very least it seems when specifying `{ upsert: false }`, the `$setOnInsert` option can be skipped completely. Indeed if the `$setOnInsert` object is empty, it probably shouldn't be included at all.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

Node.js: 12.4.0
Mongoose: 5.6.0
MongoDB: 3.6.8

**Comment by andrewpile** [comment]
Actually, it appears this is related to *renaming* `createdAt`, closing and will make new ticket.

**Comment by andrewpile** [comment]
I closed this temporarily thinking perhaps it actually only applied `createdAt` timestamps with reassigned field names. It does not. Marking `createdAt` immutable, regardless of field name, breaks `update` calls.

**Comment by vkarpov15** [comment]
Not sure how we didn't notice this issue before, but it is fixed in master and the fix will be in 5.6.2 within the next 48 hours :+1:
